Optimize tickless tick computation

c1728p9 opened the pull request:
### Description

Optimize the tick computation by computing the elapsed ticks based on relative time rather than absolute time. On the NUCLEO_L073RZ this reduces the execution time of SysTimer::resume from >35us to less than ~15us~ 20us.

~Also, to ensure the relative time calculations do not overflow set the max time the OS can suspend to 1 hour.~

### Pull request type

    [x ] Fix
    [ ] Refactor
    [ ] Target update
    [ ] Functionality change
    [ ] Docs update
    [ ] Test update
    [ ] Breaking change

### Reviewers

@pan-

c1728p9 committed:
Optimize tickless tick computation

Optimize the tick computation in the following ways:
1. Use relative time rather than absolute time
2. Replace multiplication/division pair with just multiplication
    or division:
    "* 1000000ULL / OS_TICK_FREQ"   ->   "* US_IN_TICK"
    "* OS_TICK_FREQ / 1000000"      ->   "/ US_IN_TICK"

kjbracey commented on rtos/TARGET_CORTEX/SysTimer.cpp:
```diff
@@ -106,16 +114,19 @@ uint32_t SysTimer::resume()
 
-    uint64_t new_tick = (ticker_read_us(_ticker_data) - _start_time) * OS_TICK_FREQ / 1000000;
-    if (new_tick > _tick) {
+    const timestamp_t elpased_us = ticker_read_us(_ticker_data) - _time_us;
```
`elpased`->`elapsed`

Is it worth sticking an explicit cast in for the narrowing here?

It's probably worth sticking a comment here that the extra complexity is being employed to avoid the 64-bit division in a critical section, lest anyone touch it in future.

kjbracey reviewed (commented):
This saves a fair bit on the M0, where it converts a 64-bit software division to a 32-bit software division, but on M3 and M4 it should save a *lot* more - it should be optimised into a "multiply by reciprocal" using UMULL. (I know GCC and ARMC5 do this, hopefully IAR and ARMC6 do too).

This is slightly at odds with something I'm looking to do with SysTimer - that was actually going to make it _more_ 64-bit, to help support a manual suspend API, where I was anticipating long sleep times. But I think I could deal with that by having conditional behaviour - switch between optimised 32-bit and heavy 64-bit computations depending on whether deep sleep is locked. If deep sleep is permitted, we have a free reign on interrupt latency.

c1728p9 commented on rtos/TARGET_CORTEX/SysTimer.cpp:
```diff
@@ -106,16 +114,20 @@ uint32_t SysTimer::resume()
+    // 32 bit division is used here for better performance since this is in a critical section.
+    const timestamp_t elapsed_us = (timestamp_t)(ticker_read_us(_ticker_data) - _time_us);
+    uint32_t elapsed_ticks = (elapsed_us + _remainder_us) / US_IN_TICK;
+    _remainder_us = elapsed_us - elapsed_ticks * US_IN_TICK;
```
I think this remainder calculation is wrong.

c1728p9 committed:
Optimize tickless interrupt latency

Remove unnecessary critical sections from the SysTimer code since
the access should already be serialized.

kjbracey commented on rtos/TARGET_CORTEX/SysTimer.cpp:
```diff
@@ -100,43 +102,40 @@ bool SysTimer::suspend_time_passed()
 
-    uint64_t new_tick = (ticker_read_us(_ticker_data) - _start_time) * OS_TICK_FREQ / 1000000;
-    if (new_tick > _tick) {
+    uint64_t elapsed_ticks = (ticker_read_us(_ticker_data) - _time_us) / US_IN_TICK;
```
Gah!!!  I briefly wondered whether this might be a solution, but immediately shot it down because I assumed `SysTimer::resume` was being called in a critical section from outside.  This is great.

I think it's still worth optimising for the common short case - if this is taking 20us every millisecond, that is 2% of CPU time just for the division. (On a 4MHz system I saw about 10% of CPU time being eaten up by the millisecond tickers.).

How about:

    uint64_t elapsed_us = ticker_read_us(_ticker_data) - _time_us;
    uint64_t elapsed_ticks;
    if (elapsed_us > 0xFFFFFFFF) {
        // This will typically invoke full 64/64 software division
        elapsed_ticks = elapsed_us / US_IN_TICK;
    } else {
        // This will will invoke 32/32 software division on ARMv6-M or ARMv8-M baseline, and
        // on ARMv7-M or ARMv8-M mainline can be optimised to a UMULL by 2^n/US_IN_TICK, or
        // at worst be a UDIV instruction. Thus it will be faster, maybe greatly.
        elapsed_ticks = (uint32_t) elapsed_us / US_IN_TICK;
    }

c1728p9 replied:
> I assumed `SysTimer::resume` was being called in a critical section from outside.

It used to be in a critical section, but had to be changed when the RTX API it relied on became unavailable. I missed it at this time as well.

Initially I had it avoid division all together in the fast case - when only a single tick had elapsed: 
```C
if (elapsed_time < US_IN_TICK) {
    elapsed_ticks = 0;
} else if (elapsed_time < US_IN_TICK * 2) {
    elapsed_ticks = 1;
} else {
    elapsed_ticks = elapsed_time / US_IN_TICK;
}
```

I ended up going with the simpler implementation you see now since I was just worried about reducing worst case execution time. Also, I was only setup to profile only worst-case execution time, which did not change much from this.

It sounds like you have profiled the average execution time on a couple of devices. Would you be able to profile the different optimizations and create a PR for the one which looks the most promising?

kjbracey replied:
I'll have to see if those extra cases help too. My feeling is that division tends to have early termination, so might not help much, but worth checking.

> I was just worried about reducing worst case execution time.

But we don't care about worst-case any more if we're outside of the critical section, right?

c1728p9 replied:
> But we don't care about worst-case any more if we're outside of the critical section, right?

True, worst case doesn't matter anymore. At the time I was only setup to profile worst-case, so I wasn't able to evaluate the impact of this change.

kjbracey commented on rtos/TARGET_CORTEX/SysTimer.cpp:
```diff
@@ -37,6 +37,9 @@ extern "C" {
 }
 
+#define US_IN_TICK          (1000000 / (OS_TICK_FREQ))
+MBED_STATIC_ASSERT(1000000 == US_IN_TICK * OS_TICK_FREQ, "OS_TICK_FREQ must be a divisor of 1000000 for correct tick calculations");
```
I'd have phrased that `1000000 % OS_TICK_FREQ == 0`, but whatever.

However, there's an inconsistency here - you shouldn't need the `()` around OS_TICK_FREQ above, but if you do, you'd need them here too.

c1728p9 replied:
done
